### PR TITLE
#168781683 user profile setting

### PR DESCRIPTION
**Pull request by ngireric123**
[finishes #168781683]

#### What does this PR do?
allow user to edit his/her profile settings
#### Description of Task to be completed?
- add endpoint for update user settings
#### How should this be manually tested?
- Clone the repository from [here](https://github.com/andela/caret-bn-backend.git)
- Go to the repo directory
- Go to Ft-profile-setting-168781683 branch
- Run `npm install`
- Go to the following url https://developers.google.com/oauthplayground/ and authorize the following APIs (This is done in step 1 of the Playground process).
 a) Google OAuth apiv2
      - https://www.googleapis.com/auth/userinfo.email
      - https://www.googleapis.com/auth/userinfo.profile
- Exchange the authorization code for tokens (This is done in step 2 of the Playground process)
- Copy the new access token for your `GOOGLE_ACCESS_TOKEN=` in your .env file

- run `sequelize db:migrate`
- run `sequelize db:seed:all`
- run `npm run dev`
 **Edit profile**
- In Postman Provide a login bearer token
- Fill the body with all data you want to edit
- send a `PATCH` request to `http://localhost:<patch>/api/v1/users/{email}`

#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
https://www.pivotaltracker.com/story/show/168781683
#### Screenshots (if appropriate)
image
image

#### Questions:
N/A

### CODE REVIEW

File: src/database/migrations/20190930143254-User.js [hunk 1]
```diff
@@ -18,6 +18,47 @@ module.exports = {
         }
       }
     },
+    // role: {
+    //   type: Sequelize.INTEGER,
+    //   unique: false,
+    //   defaultValue: 1
+    // },
```
Comment (alainmateso) [inline src/database/migrations/20190930143254-User.js]: Please remove commented code!

**Review by nignanthomas — CHANGES_REQUESTED** [review]
Great job @ngireric123. I tried it and it seems to work fine.
However, I would like you to add a few more tests, like
- when the ***token is malformed**
- when there is **no token**
- when the **payload email !== req.params.email**
Those lines seem to not be covered by tests.

File: src/controllers/profile.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+      );
+      const {
+        password, createdAt, updatedAt, isVerified, ...data
+      } = newUser[1].dataValues;
+      return responseUtil(res, 200, string.users.success.SUCCESS_UPDATE, data);
+    } catch (error) {
+      return responseError(res, 500, string.users.error.SOMETHING_WRONG);
+    }
+  }
+}
```
Comment (alainmateso) [inline src/controllers/profile.js]: Please respect naming conventions when naming files. I think this file should be profileController.js

**Review by alainmateso — CHANGES_REQUESTED** [review]
You have done a good job Eric! However, there are some changes I would like you to implement like naming conventions, removing commented code, removing duplicated files or having multiple files that do the same thing(Files that sign the token, files that decode the token...). Kindly look into that!

**Review by alainmateso — APPROVED** [review]
Awesome

**Review by sabin18 — APPROVED** [review]
good job @ngireric123

**Review by nignanthomas — APPROVED** [review]

File: src/controllers/profileController.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    try {
+      const newUser = await users.update(
+        req.body,
+        {
+          where: { email: req.params.email }, returning: true, plain: true
+        }
+      );
+      const {
+        password, createdAt, updatedAt, isVerified, ...data
+      } = newUser[1].dataValues;
```
Comment (nakiwuge) [inline src/controllers/profileController.js:25]: Add a newline here

File: src/controllers/profileController.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+  static async getProfile(req, res) {
+    if (req.user.payload.email !== req.params.email) {
+      return responseError(res, 400, string.users.error.USER_SAME_EMAIL);
+    }
+    try {
+      const { email } = req.params;
+      const user = await users.findOne({
+        where: { email }, raw: true
+      });
```
Comment (nakiwuge) [inline src/controllers/profileController.js:38]: Add a new line here

File: src/controllers/profileController.js [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+      if (!user) {
+        return res.status(404).json({ error: 'user does not exist' });
+      }
```
Comment (nakiwuge) [inline src/controllers/profileController.js:43]: Add a newline here

File: src/routes/api/users.js [hunk 6]
```diff
@@ -198,11 +202,60 @@ const router = express.Router();
...
+ *         description: interrnal server error
+ *       '400':
+ *         description: invalid data
+ */
 
 router.post('/register', validateSignup, verifyExist, confirmPassword, signup);
 router.get('/verify/:token', UserController.userVerify);
 router.post('/forgotpassword', validateResetpassword.checkEmail, UserController.Providelink);
 router.patch('/resetpassword/:token', EmailToken.UseraccessRequired, validateResetpassword.checkReset, UserController.Changepassword);
+router.patch('/:email', validateToken, validateProfile, updateProfile);
```
Comment (nakiwuge) [inline src/routes/api/users.js]: this route should also be `/profile/:email`

**Review by nakiwuge — CHANGES_REQUESTED** [review]
@ngireric123 good job with the implementation. I have left a few comments kindly check them out.

**Review by nakiwuge — APPROVED** [review]

**Pull request merged**